src/shaders/atmosphere.fragment.glsl

```diff
@@ -0,0 +1,136 @@
+uniform float u_transition;
+uniform highp float u_fadeout_range;
+uniform highp float u_temporal_offset;
+uniform vec3 u_start_color;
+uniform vec4 u_color;
+uniform vec4 u_space_color;
+uniform vec4 u_high_color;
+uniform float u_star_intensity;
+uniform float u_star_size;
+uniform float u_star_density;
+uniform float u_horizon_angle;
+uniform mat4 u_rotation_matrix;
+
+varying highp vec3 v_ray_dir;
+varying highp vec3 v_horizon_dir;
+
+float random(vec3 p) {
+    p = fract(p * vec3(23.2342, 97.1231, 91.2342));
+    p += dot(p.zxy, p.yxz + 123.1234);
+    return fract(p.x * p.y);
+}
+
+float stars(vec3 p, float scale, vec2 offset) {
+    vec2 uv_scale = (u_viewport / u_star_size) * scale;
+    vec3 position = vec3(p.xy * uv_scale + offset * u_viewport, p.z);
+
+    vec3 q = fract(position) - 0.5;
+    vec3 id = floor(position);
+
+    float random_visibility = step(random(id), u_star_density);
+    float circle = smoothstep(0.5 + u_star_intensity, 0.5, length(q));
+
+    return circle * random_visibility;
+}
+
+void main() {
+    highp vec3 dir = normalize(v_ray_dir);
+
+#ifdef PROJECTION_GLOBE_VIEW
+    float globe_pos_dot_dir = dot(u_globe_pos, dir);
+    highp vec3 closest_point_forward = abs(globe_pos_dot_dir) * dir;
+    float norm_dist_from_center = length(closest_point_forward - u_globe_pos) / u_globe_radius;
+
+    if (norm_dist_from_center < 1.0) {
+        gl_FragColor = u_color;
+    }
+#endif
+
+    highp vec3 horizon_dir = normalize(v_horizon_dir);
+    float horizon_angle_mercator = dir.y < horizon_dir.y ?
+        0.0 : max(acos(dot(dir, horizon_dir)), 0.0);
+
+#ifdef PROJECTION_GLOBE_VIEW
+    // Angle between dir and globe center
+    highp vec3 closest_point = globe_pos_dot_dir * dir;
+    float closest_point_to_center = length(closest_point - u_globe_pos);
+    float theta = asin(clamp(closest_point_to_center / length(u_globe_pos), -1.0, 1.0));
+
+    // Backward facing closest point rays should be treated separately
+    float horizon_angle = globe_pos_dot_dir < 0.0 ?
+        PI - theta - u_horizon_angle : theta - u_horizon_angle;
+
+    // Increase speed of change of the angle interpolation for
+    // a smoother visual transition between horizon angle mixing
+    float angle_t = pow(u_transition, 10.0);
+
+    horizon_angle = mix(horizon_angle, horizon_angle_mercator, angle_t);
+#else
+    float horizon_angle = horizon_angle_mercator;
+#endif
+
+    // Normalize in [0, 1]
+    horizon_angle /= PI;
+
+    // exponential curve
+    // [0.0, 1.0] == inside the globe, > 1.0 == outside of the globe
+    // https://www.desmos.com/calculator/l5v8lw9zby
+    float t = exp(-horizon_angle * pow(u_fadeout_range, -1.0));
```

Comment:
is this an optimization to write `pow(x, -1)` instead of `1/x`?

Reply:
Not an optimization, that's an overlook from converting the function from desmos, thanks for pointing that out I'll update it!

Reply:
Also, I switched that function to prefer `exp` over an `sqrt` based function, so that the effect is smoother, you can notice that the `sqrt` based function that we used in `main` (in red) results in an abrupt change when crossing `y=0` due to clamping:

![Screenshot from 2022-04-13 11-29-45](https://user-images.githubusercontent.com/7061573/163246621-65cbb24a-0dbd-42d1-989c-8acae2f2e858.png)

Reply:
yes I noticed! That is a better option :+1: